VPLAY-10193 pssh parsing for mp4demux

pstroffolino opened the pull request:
VPLAY-10193 pssh parsing for mp4demux

support for extracting pssh and encryption metadata from encrypted mp4 content

pstroffolino committed:
VPLAY-10193 pssh parsing for mp4demux

Reason for Change: encryption support in mp4demux
Test Guidance: enable mp4demux as alternative to qtdemux and play encrypted content
Risk: Low (not used by default)

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
mend

pstroffolino committed:
VPLAY-10193 mp4demux support for parsing protection data

Reason for Change: changes to extract pssh header and drm metadata
Test Guidance: refer ticket
Risks: Low

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
VPLAY-10193 drm support for mp4demux

Reason for Change: follow-up improvements
Test Guidance: Refer Ticket
Risk: low (not yet enabled by default)

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
VPLAY-10193 cleanup - l1 test fix

Reason for Change: missing include <string>
Test Guidance: l1 tests passing
Risk: Low

pstroffolino committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10193_pssh

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
feb329f2a737cf429e7cd791558ae5a4
PRESENT
-->

# Coverity Issue - Variable copied when it could be moved
"sample" is copied and then passed-by-reference as parameter to STL insertion function "std::vector<Mp4Sample, std::allocator<Mp4Sample> >::push_back(std::vector<Mp4Sample, std::allocator<Mp4Sample> >::value_type const &)", when it could be moved instead.

_Low Impact, CWE-none_
_COPY_INSTEAD_OF_MOVE_

## How to fix
 Use "std::move(""sample"")" instead of "sample".

## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[test/gstTestHarness/mp4demux.hpp:469](undefined//blob/undefined/test/gstTestHarness/mp4demux.hpp#L469)

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
a9052a6ccdb80732e7a8e0a67bd6fed6
PRESENT
-->

# Coverity Issue - Out-of-bounds read
Overrunning array "timescale" of 2 4-byte elements at element index 2 (byte offset 11) using index "mediaType" (which evaluates to 2).

_High Impact, CWE-125_
_OVERRUN_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[middleware/InterfacePlayerRDK.cpp:2931](undefined//blob/undefined/middleware/InterfacePlayerRDK.cpp#L2931)

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
eeb49df7bb16cd4c81af4a79c4075715
PRESENT
-->

# Coverity Issue - Unchecked return value
Calling "gst_element_seek" without checking return value (as is done elsewhere 6 out of 6 times).

_Medium Impact, CWE-252_
_CHECKED_RETURN_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[test/gstTestHarness/gst-port.cpp:406](undefined//blob/undefined/test/gstTestHarness/gst-port.cpp#L406)

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
c4a59ffb8ad08c1f9475e6d500b4be9c
PRESENT
-->

# Coverity Issue - Parse warning
declaration hides parameter "len" (declared at line 2799)

_Low Impact, CWE-398_
_PW.PARAMETER_HIDDEN_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[middleware/InterfacePlayerRDK.cpp:2937](undefined//blob/undefined/middleware/InterfacePlayerRDK.cpp#L2937)

rdkcmf-jenkins commented:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
0010245a64dad84e993c6f5132055bf6
PRESENT
-->

# Coverity Issue - Logically dead code
Execution cannot reach the expression """" inside this statement: "g_print("Pipeline::SendBuff...".

_Medium Impact, CWE-561_
_DEADCODE_



## Issue location
This issue was discovered outside the diff for this Pull Request. You can find it at:
[test/gstTestHarness/gst-port.cpp:373](undefined//blob/undefined/test/gstTestHarness/gst-port.cpp#L373)

rdkcmf-jenkins commented on middleware/InterfacePlayerRDK.cpp:
```diff
@@ -2961,7 +2961,7 @@ bool InterfacePlayerRDK::SendHelper(int type, const void *ptr, size_t len, doubl
 				else
 				{ // init header
-					timescale[mediaType] = mp4Demux->timescale;
+					timescale[mediaType] = mp4Demux->getTimeScale();
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
a9052a6ccdb80732e7a8e0a67bd6fed6
PRESENT
-->

# Coverity Issue - Out-of-bounds write
Overrunning array "timescale" of 2 4-byte elements at element index 2 (byte offset 11) using index "mediaType" (which evaluates to 2).

_High Impact, CWE-119_
_OVERRUN_

rdkcmf-jenkins replied:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
a9052a6ccdb80732e7a8e0a67bd6fed6
PRESENT
-->

# Coverity Issue - Out-of-bounds write
Overrunning array "timescale" of 2 4-byte elements at element index 2 (byte offset 11) using index "mediaType" (which evaluates to 2).

_High Impact, CWE-119_
_OVERRUN_

pstroffolino committed:
VPLAY-10193 mp4demux refactoring

Reason for Change: code cleanup
Test Guidance: gst test harness and aamp playback (latter with useMp4Demux)
Risks: Low

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

rdkcmf-jenkins commented on middleware/InterfacePlayerRDK.cpp:
```diff
@@ -2926,9 +2926,14 @@ bool InterfacePlayerRDK::SendHelper(int type, const void *ptr, size_t len, doubl
-				// some lldash streams don't have timescale in media segments
-				Mp4Demux *mp4Demux = new Mp4Demux(ptr,len,timescale[mediaType]);
+				static Mp4Demux *m_mp4Demux[2];
+				Mp4Demux *mp4Demux = m_mp4Demux[mediaType];
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
fe55d288f9000509728467c624b551e3
PRESENT
-->

# Coverity Issue - Out-of-bounds read
Overrunning array "m_mp4Demux" of 2 8-byte elements at element index 2 (byte offset 23) using index "mediaType" (which evaluates to 2).

_High Impact, CWE-125_
_OVERRUN_

rdkcmf-jenkins commented on middleware/InterfacePlayerRDK.cpp:
```diff
@@ -2926,9 +2926,14 @@ bool InterfacePlayerRDK::SendHelper(int type, const void *ptr, size_t len, doubl
+				if( !mp4Demux )
+				{
+					mp4Demux = new Mp4Demux();
+					m_mp4Demux[mediaType] = mp4Demux;
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
fe55d288f9000509728467c624b551e3
PRESENT
-->

# Coverity Issue - Out-of-bounds write
Overrunning array "m_mp4Demux" of 2 8-byte elements at element index 2 (byte offset 23) using index "mediaType" (which evaluates to 2).

_High Impact, CWE-119_
_OVERRUN_

pstroffolino committed:
VPLAY-10193 fix for dai2 test, moved mp4demux.cpp into mp4demux.hpp

Reason for Change: add'l cleanup and gst test harness regression
Risk: Low

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
VPLAY-10193 AdjustMediaDecodeTime optimization

Reason for Change: follow-up code cleanup

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
VPLAY-10193 fixes for GstStructure metadata crash

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
VPLAY-10193 add'l comments in mp4demux parser

Reason for Change: added indent cap (buffer guard)  and more documentation

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10193_pssh

pstroffolino committed:
added bounds check flagged by coverity

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

rdkcmf-jenkins commented on test/gstTestHarness/gst-test.cpp:
```diff
@@ -194,8 +188,11 @@ class TrackFragment: public TrackEvent
+				{
+					gMp4Demux[mediaType] = new Mp4Demux();
+				}
+				gMp4Demux[mediaType]->Parse(ptr,len);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
6544e5df56acd0c74e03653967055800
PRESENT
-->

# Coverity Issue - Untrusted loop bound
Passing tainted expression "*this->ptr" to "Parse", which uses it as a loop boundary.

_Medium Impact, CWE-606_
_TAINTED_SCALAR_

## How to fix
 Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.

rdkcmf-jenkins replied:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
6544e5df56acd0c74e03653967055800
PRESENT
-->

# Coverity Issue - Untrusted loop bound
Passing tainted expression "*this->ptr" to "Parse", which uses it as a loop boundary.

_Medium Impact, CWE-606_
_TAINTED_SCALAR_

## How to fix
 Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.

pstroffolino committed:
added new 'parse' command in aampcli

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
VPLAY-10193 misc coverity fixes

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10193_pssh

rdkcmf-jenkins commented on test/aampcli/AampcliPlaybackCommand.cpp:
```diff
@@ -1067,6 +1072,50 @@ void PlaybackCommand::addCommand(std::string command,std::string description)
+					if( rc == len )
+					{
+						auto mp4Demux = new Mp4Demux(true);
+						mp4Demux->Parse(ptr,len);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
4c001665fc4445991258e544236be55b
PRESENT
-->

# Coverity Issue - Untrusted loop bound
Passing tainted expression "*ptr" to "Parse", which uses it as a loop boundary.

_Medium Impact, CWE-606_
_TAINTED_SCALAR_

## How to fix
 Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.

rdkcmf-jenkins commented on test/gstTestHarness/gst-test.cpp:
```diff
@@ -175,56 +172,59 @@ void GetVideoSegmentPath( char path[MAX_PATH_SIZE], int segmentNumber, VideoReso
+					{
+						gMp4Demux[mediaType] = new Mp4Demux();
+					}
+					gMp4Demux[mediaType]->Parse(ptr,len);
```
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
6544e5df56acd0c74e03653967055800
PRESENT
-->

# Coverity Issue - Untrusted loop bound
Passing tainted expression "*this->ptr" to "Parse", which uses it as a loop boundary.

_Medium Impact, CWE-606_
_TAINTED_SCALAR_

## How to fix
 Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.

rdkcmf-jenkins replied:
<!-- Comment managed by coverity-report-output-v7 action, do not modify!
6544e5df56acd0c74e03653967055800
PRESENT
-->

# Coverity Issue - Untrusted value as argument
Passing tainted expression "*this->ptr" to "Parse", which uses it as an offset.

_Medium Impact, CWE-20_
_TAINTED_SCALAR_

## How to fix
 Ensure that tainted values are properly sanitized, by checking that their values are within a permissible range.

pstroffolino committed:
coverity directive

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
Merge branch 'feature/VPLAY-10193_pssh' of https://github.com/rdkcentral/aamp into feature/VPLAY-10193_pssh

pstroffolino committed:
pretty formatting

Signed-off-by: Stroffolino <philip_stroffolino@cable.comcast.com>

pstroffolino committed:
Merge branch 'dev_sprint_25_2' into feature/VPLAY-10193_pssh
